feat: Adds support for STS response not returning expires_in field.

xil222 opened the pull request:
stsResponse doesnt return expire_in if subject_token isnt service account access token. 
Change the expire_in field to optional and add changes to downscopedclient and baseexternalclient.
bug: b/193924974
Your feedback is greatly appreciated! @bojeil-google @bcoe

xil222 committed:
Adds support for stsResponse doesnt return expiry time.

xil222 committed:
Adds support for STS response not returning expires_in field.

xil222 committed:
Adds support for STS response not returning expires_in field.

xil222 committed:
Adds support for stsResponse doesnt return expiry time.

xil222 committed:
Merge branch 'expireTime' of https://github.com/xil222/google-auth-library-nodejs into expireTime

Commit:
Merge branch 'master' into expireTime

Commit:
meta: add nodejs-auth as codeowner (#1217)

averikitsch committed:
samples: update TODO section (#1218)

* fix: update TODO section

* remove origin

* Update idtokens-serverless.js

lsirac commented on src/auth/downscopedclient.ts:
```diff
@@ -214,10 +214,15 @@ export class DownscopedClient extends AuthClient {
       this.credentialAccessBoundary
     );
 
+    const sourceCredExpireDate =
```
Let's add a comment here: 

The STS endpoint will only return the expiration time for the downscoped access token if the
original access token represents a service account. The downscoped token's expiration time will always match the source credential expiration. When no expires_in is returned, we can copy the source credential's expiration time.

lsirac commented on test/test.baseexternalclient.ts:
```diff
@@ -1053,6 +1053,74 @@ describe('BaseExternalAccountClient', () => {
         scopes.forEach(scope => scope.done());
       });
 
+      it('should return credential with no expiry date if STS response does not return one', async () => {
```
I think expires_in is always returned for external account credentials, so this test might not be needed/confusing.

xil222 replied:
Yes, I think this usually not needed. However, since I changed logic by setting expire_in to optional,
I want to cover the line expireDate = null in refreshAccessTokenAsync() else case in baseexternalclient.ts.

lsirac commented on test/test.downscopedclient.ts:
```diff
@@ -374,6 +374,126 @@ describe('DownscopedClient', () => {
+      scope.done();
+    });
+
+    it('should has no expiry date if source cred has no expiry time and STS response does not return one', async () => {
```
nit: have

lsirac commented on test/test.downscopedclient.ts:
```diff
@@ -374,6 +374,126 @@ describe('DownscopedClient', () => {
       scope.done();
     });
+
+    it('should copy source cred expiry time if STS response does not return expiry time', async () => {
```
This test confuses me. Based on the description, all we should be doing is (which you are doing):
1. Initialize a source credential with a certain expiration time
2. Mock STS to not return expires_in
3. Initialize DownscopedClient with the source credential
4. Call getAccessToken, and assert that the expiration matches the source credential

I don't think you need mock the clock so that the credentials you set expire. If you want to test that path, that can be a separate test?

Commit:
feat(impersonated): add impersonated credentials auth (#1207)

release-please[bot] committed:
chore: release 7.4.0 (#1220)

:robot: I have created a release \*beep\* \*boop\*
---
## [7.4.0](https://www.github.com/googleapis/google-auth-library-nodejs/compare/v7.3.0...v7.4.0) (2021-07-29)


### Features

* **impersonated:** add impersonated credentials auth ([#1207](https://www.github.com/googleapis/google-auth-library-nodejs/issues/1207)) ([ab1cd31](https://www.github.com/googleapis/google-auth-library-nodejs/commit/ab1cd31e07d45424f614e0401d1068df2fbd914c))
---


This PR was generated with [Release Please](https://github.com/googleapis/release-please). See [documentation](https://github.com/googleapis/release-please#release-please).

xil222 committed:
fix(downscoped-client): bug fixes for downscoped client implementation. (#1219)

release-please[bot] committed:
chore: release 7.4.1 (#1222)

:robot: I have created a release \*beep\* \*boop\*
---
### [7.4.1](https://www.github.com/googleapis/google-auth-library-nodejs/compare/v7.4.0...v7.4.1) (2021-07-29)


### Bug Fixes

* **downscoped-client:** bug fixes for downscoped client implementation. ([#1219](https://www.github.com/googleapis/google-auth-library-nodejs/issues/1219)) ([4fbe67e](https://www.github.com/googleapis/google-auth-library-nodejs/commit/4fbe67e08bce3f31193d3bb7b93c4cc1251e66a2))
---


This PR was generated with [Release Please](https://github.com/googleapis/release-please). See [documentation](https://github.com/googleapis/release-please#release-please).

averikitsch committed:
sample: update region tag (#1223)

xil222 committed:
Remove extra tests in baseExternalClient.

xil222 committed:
merge origin/expireTime

xil222 committed:
Merge branch 'master' into expireTime

bcoe commented on src/auth/baseexternalclient.ts:
```diff
@@ -378,10 +378,11 @@ export abstract class BaseExternalAccountClient extends AuthClient {
         stsResponse.access_token
       );
     } else {
```
If you made this `else if (stsResponse.expires_in) {`

Then there's no need to cast to a `number`.

lsirac commented on test/test.downscopedclient.ts:
```diff
@@ -411,9 +411,11 @@ describe('DownscopedClient', () => {
       clock.tick(1);
       const refreshedTokenResponse = await downscopedClient.getAccessToken();
+
+      const responseExpireIn = stsSuccessfulResponse.expires_in as number;
```
nit: responseExpiresIn or expiresIn

xil222 committed:
change parsing expire_in logic in baseexternalclient.

xil222 committed:
Merge branch 'expireTime' of https://github.com/xil222/google-auth-library-nodejs into expireTime

xil222 committed:
Merge branch 'master' into expireTime

bojeil-google commented on test/test.downscopedclient.ts:
```diff
@@ -534,12 +536,100 @@ describe('DownscopedClient', () => {
+
+    it('should copy source cred expiry time if STS response does not return expiry time', async () => {
+      const now = new Date().getTime();
+      const credentials = {
+        access_token: 'ACCESS_TOKEN',
+        expiry_date: now + ONE_HOUR_IN_SECS * 1000,
+      };
```
Why do you need to set these credentials and to call `client.setCredentials(credentials);`?
I suggest removing them for this test to just focus on the expiration time population logic.

Instead you can expose a public property on `TestAuthClient` to populate the expiration time on `client.credentials.expiry_date` when `getAccessToken()` is called.

xil222 replied:
I think there're some confusions here, getAccessToken() is contained both in downscopedClient and authClient however functions differently, in testAuthClient, it is only supposed to return the access token / subject token, the latter one is returning an object including(token, expirationTime, res). I am mocking testAuthClient class extends on authClient class but not downscopedClient.

In implementation, the actual way of getting sourceCredExpireDate() is this.authClient.credentials?.expiry_date but not calling getAccessToken(), so calling client.setCredential(credentials) is the only and built-in way of setting expiration time.

bojeil-google commented on test/test.downscopedclient.ts:
```diff
@@ -534,12 +536,100 @@ describe('DownscopedClient', () => {
+            subject_token: 'subject_token_0',
+            subject_token_type: 'urn:ietf:params:oauth:token-type:access_token',
+            options:
+              testClientAccessBoundary &&
```
You don't need this check.

bojeil-google commented on test/test.downscopedclient.ts:
```diff
@@ -534,12 +536,100 @@ describe('DownscopedClient', () => {
+            subject_token: 'subject_token_0',
+            subject_token_type: 'urn:ietf:params:oauth:token-type:access_token',
+            options:
+              testClientAccessBoundary &&
```
You don't need this check.

bojeil-google commented on test/test.downscopedclient.ts:
```diff
@@ -534,12 +536,100 @@ describe('DownscopedClient', () => {
+      };
+      const stsSuccessfulResponseWithoutExpireInField = Object.assign(
+        stsSuccessfulResponse,
+        {}
```
Same issue here. The arguments need to be swapped.

bojeil-google commented on test/test.downscopedclient.ts:
```diff
@@ -534,12 +536,100 @@ describe('DownscopedClient', () => {
+    });
+
+    it('should have no expiry date if source cred has no expiry time and STS response does not return one', async () => {
+      const credentials = {
+        access_token: 'ACCESS_TOKEN',
+      };
```
I think `TestAuthClient ` already doesn't populate the expiration, so this may not be needed.

bojeil-google reviewed (changes requested):
Thanks for putting this together. I mostly have suggestions and propose tweaks to the unit tests.

xil222 committed:
Add testAuthClient setter and tweak some tests.

xil222 committed:
Merge branch 'expireTime' of https://github.com/xil222/google-auth-library-nodejs into expireTime

bojeil-google commented on test/test.downscopedclient.ts:
```diff
@@ -49,6 +49,10 @@ class TestAuthClient extends AuthClient {
     throw new Error('Cannot get subject token.');
   }
 
+  set expirationTime(expirationTime: number) {
```
I think this could also be cleared or set to null?
`set expirationTime(expirationTime: number|undefined|null) {`

bojeil-google commented on test/test.baseexternalclient.ts:
```diff
@@ -356,14 +357,34 @@ describe('BaseExternalAccountClient', () => {
         );
+        // Listen to tokens events. On every event, push to list of
+        // emittedEvents.
+        client.on('tokens', tokens => {
```
BTW, should we add a similar test for downscoped client to assert emitted event?

xil222 committed:
add emit events for some tests.

Commit:
Merge branch 'master' into expireTime
